openghg/analyse/_scenario.py

```diff
@@ -1287,102 +1285,78 @@ def _calc_modelled_obs_HiTRes(
+        def compute_fp_x_flux(
+            fp_HiTRes: xr.DataArray,
+            flux_high_freq: xr.DataArray,
+            flux_low_freq: xr.DataArray,
+            highest_res_H: int,
+            max_h_back: int,
+        ) -> xr.DataArray:
 
-            # convert to array to use in numba loop
-            flux_high_freq = da.array(flux_high_freq)
-            flux_low_freq = da.array(flux_low_freq)
+            # do low res calculation
+            fp_residual = fp_HiTRes.sel(H_back=fp_HiTRes.H_back.max(), drop=True)  # take last H_back value
+            flux_low_freq = flux_low_freq.reindex_like(fp_residual, method="ffill")  # forward fill times
 
-            # Multiply the HiTRes footprints with the HiTRes emissions to give mf
-            # Multiply residual footprints by low frequency emissions data to give residual mf
-            # flux units : mol/m2/s;       fp units : mol/mol/mol/m2/s
-            # --> mol/mol/mol/m2/s * mol/m2/s === mol / mol
-            fpXflux_time = flux_high_freq * fp_high_freq
             fpXflux_residual = flux_low_freq * fp_residual
 
-            # append the residual emissions
-            fpXflux_time = np.dstack((fpXflux_time, fpXflux_residual))
-
-            if output_fpXflux:
-                # Sum over time (H back) to give the total mf at this timestep
-                fpXflux[:, :, tt] = fpXflux_time.sum(axis=2)
+            # get high freq fp
+            fp_high_freq = fp_HiTRes.where(fp_HiTRes.H_back != fp_HiTRes.H_back.max(), drop=True)
 
-            if output_TS:
-                # work out timeseries by summing over lat, lon (24 hrs)
-                timeseries[tt] = fpXflux_time.sum()
+            # if flux_res_H > 24, then flux_high_freq = flux_low_freq, and we don't take a sum over windows of flux_high_freq
+            if flux_res_H > 24:
+                fpXflux = (flux_low_freq * fp_high_freq).sum("H_back")
+            else:
+                flux_high_freq = make_hf_flux_rolling_avg_array(
+                    flux_high_freq, fp_high_freq, highest_res_H, max_h_back
+                )
+                fpXflux = (flux_high_freq * fp_high_freq).sum("H_back")
 
-        # TODO: Add details about units to output
+            return fpXflux + fpXflux_residual
 
-        if output_fpXflux:
-            fpXflux = DataArray(
-                fpXflux,
-                dims=("lat", "lon", "time"),
-                coords={"lat": lat, "lon": lon, "time": time_array},
-            )
+        fpXflux = compute_fp_x_flux(
+            fp_HiTRes,
+            flux_high_freq,
+            flux_low_freq,
+            highest_res_H,
+            max_h_back,
+        )
 
         if output_TS:
-            timeseries = DataArray(timeseries, dims=("time"), coords={"time": time_array})
+            timeseries = fpXflux.sum(["lat", "lon"])
+
+        # TODO: Add details about units to output
 
         if output_fpXflux and output_TS:
             timeseries.compute()
```

Comment:
We might not want to compute the `fpXflux` here since it will be used for later calculations in inversions (once inversions switches to using this code to get `fpXflux`) and it would probably be best to delay computation as long as possible.

Reply:
Did you want to add an issue for this?

Reply:
I don't think it's possible to get `fpXflux` without calling this "private" method directly right now. If we don't have an issue for that, we should make one, since this code gets duplicated in inversions for that reason. So... delaying compute could be done now, or I could mention it in that issue.

Reply:
No, `timeseries.compute()` will do that. For inversions, we don't actually need the timeseries, we need the `fpXflux` dataset. (Similarly for `bc_mod`, we actually need the dataset used to compute that.)

Reply:
If you want both `timeseries` and `fpXflux` and don't mind using the memory to store `fpXflux`, then you could use `.compute` or `.persist` on `fpXflux`.

Reply:
I don't think there was a strong reason why `fpXflux` was computed (or `timeseries` really) at that point before - if we wanted more control over that we could allow for that to be flagged and by default not computed.

Also if it would be more useful to split this function out into more steps so it could be accessed in a more fine-grained way for inversions that would be fine too.